components/raftstore/src/store/fsm/apply.rs

```diff
@@ -1652,73 +1679,145 @@ where
-                );
+                (Some(exist_peer), cct) => match (exist_peer.get_role(), cct) {
+                    (PeerRole::IncomingVoter, _) | (PeerRole::DemotingVoter, _) => {
+                        error!(
+                            "can't apply confchange because configuration is still in joint state";
+                            "region_id" => self.region_id(),
+                            "peer_id" => self.id(),
+                            "peer" => ?peer,
+                            "region" => ?&self.region
+                        );
+                        return Err(box_err!(
+                            "can't apply confchange {:?} to peer {:?} from region {:?}",
+                            cct,
+                            peer,
+                            self.region
+                        ));
+                    }
+                    // Add node
+                    (_, ConfChangeType::AddNode) | (_, ConfChangeType::AddLearnerNode) => {
+                        let (role, exist_id, incoming_id) =
+                            (exist_peer.get_role(), exist_peer.get_id(), peer.get_id());
+
+                        if exist_id != incoming_id // Add peer with different id to the same store
+                            // The peer is already the requested role
+                            || (role, cct) == (PeerRole::Voter, ConfChangeType::AddNode)
+                            || (role, cct) == (PeerRole::Learner, ConfChangeType::AddLearnerNode)
+                        {
+                            error!(
+                                "can't add duplicated peer";
+                                "region_id" => self.region_id(),
+                                "peer_id" => self.id(),
+                                "peer" => ?peer,
+                                "exist peer" => ?exist_peer,
+                                "confchnage type" => ?cct,
+                                "region" => ?&self.region
+                            );
+                            return Err(box_err!(
+                                "can't add duplicated peer {:?} to region {:?}, duplicated with exist peer {:?}",
+                                peer,
+                                self.region,
+                                exist_peer
+                            ));
+                        }
+                        match (role, cct) {
+                            (PeerRole::Voter, ConfChangeType::AddLearnerNode) => match kind {
+                                ConfChangeKind::Simple => exist_peer.set_role(PeerRole::Learner),
+                                ConfChangeKind::EnterJoint => {
+                                    exist_peer.set_role(PeerRole::DemotingVoter)
+                                }
+                                _ => unreachable!(),
+                            },
+                            (PeerRole::Learner, ConfChangeType::AddNode) => match kind {
```

Comment:
Old code will reject demoting voter and new code will accept it happily, a wrong conf change during upgrading can lead to inconsistency.

Reply:
Maybe safest way is leave the old implement untouched and handle v2 separately. This also prevent future updates from bringing corner cases.